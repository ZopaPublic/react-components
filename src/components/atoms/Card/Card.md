The design team has agreed on three types of cards depending on the context.

#### card type (_default_)

A standard or default card, which we'll assume if the type property has been omitted or set explicitly to card. As of this writing, cards of this type are non-interactable and have a border radius of 4px.

```js { "props": { "style": { "backgroundColor": "#141E64" } } }
import { Card } from '@zopauk/react-components';

<Card type="card">
  <p>I'm a card ðŸ˜€ , with an explicit type prop.</p>
</Card>;
```

#### linkCard type

A card that is meant to be clickable and respond to user interactions (hover effects et al). Border radius of 8px. Note that any additional **styles and effects are meant to be added separately** by using this component as a base on which to build another with the desired effects.

Cards of type `"linkCard"` are meant to be used to build components that interact with user actions.

```js { "props": { "style": { "backgroundColor": "#141E64" } } }
import { Card, Header3 } from '@zopauk/react-components';

<div onClick={() => alert("You're a winner!")}>
  <Card type="linkCard">
    <Header3>New rates today! ðŸ’¸</Header3>
    <p>Find out how you can benefit the most from our new rates.</p>
  </Card>
</div>;
```

#### button type

Same as above, but meant to appear as a button with minimal content. Has semicircular borders on horizontal axis.

A card is entirely white (`colors.base.white`) by default.

```jsx
<div>
  <div data-comment="Button 1" style={{ marginBottom: '20px' }}>
    <Card type="button" backgroundColor={colors.primary.teal600}>
      I look like a button
    </Card>
  </div>

  <div data-comment="Button 2" style={{ marginBottom: '20px' }}>
    <Card type="button" backgroundColor={colors.primary.teal600}>
      <div style={{ textAlign: 'center' }}>Mix and match for desired effects</div>
    </Card>
  </div>

  <div data-comment="Button 3">
    <Card type="button" backgroundColor={colors.primary.teal600} display="inline-block">
      Inline button example
    </Card>
  </div>
</div>
```
