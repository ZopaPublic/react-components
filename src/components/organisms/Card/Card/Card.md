### Summary

`<Card />` renders a div that is in charge of laying out and styling inner card components.

- It's meant to be used along with other [`<Card>`](/#/Components/Organisms/Card/Card) components.
- Check out [`<Card />`](/#/Components/Organisms/Card) examples to see how they're all meant to play together.
