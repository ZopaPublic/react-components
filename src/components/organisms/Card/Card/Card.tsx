import * as React from 'react';
import styled from 'styled-components';
import { colors } from '../../../../constants';
import { CardImageContainer } from '../CardImage/CardImage';
import CardText from '../CardText/CardText';
import CardLineItem from '../CardLineItem/CardLineItem';
import { AppTheme, useThemeContext } from '../../../styles/Theme';

export type CardLayout = 'horizontal' | 'vertical';
export type CardStyling = 'primary' | 'secondary' | 'brand' | 'action' | 'info';

export interface CardProps {
  /**
   * Determines the layout of the card
   * @default `vertical`
   */
  layout?: CardLayout;
  /**
   * Determines the styling of the card
   * @default `secondary`
   */
  styling?: CardStyling;
}

const layoutDirections: Record<CardLayout, string> = {
  horizontal: 'row',
  vertical: 'column',
};

const layoutImageStyles: Record<CardLayout, string> = {
  horizontal: '200px',
  vertical: '200px',
};

const StyledCard = styled.div<CardProps & { theme: AppTheme }>`
  height: 100%;
  width: 100%;
  display: flex;
  overflow: hidden;
  background-color: ${({ theme, styling = 'secondary' }) => theme.card[styling].backgroundStyle};
  border: ${({ theme, styling = 'secondary' }) => theme.card[styling].borderStyle};
  border-radius: ${({ theme, styling = 'secondary' }) => theme.card[styling].borderRadius};
  box-shadow: ${({ theme, styling = 'secondary' }) => theme.card[styling].boxShadowStyle};
  flex-direction: ${({ layout = 'vertical' }) => layoutDirections[layout]};
  & > ${CardImageContainer} {
    flex: 0 1 ${({ layout = 'vertical' }) => layoutImageStyles[layout]};
  }
  & .zrc__card-heading {
    font-size: ${({ theme, styling = 'secondary' }) => theme.card[styling].headingSize};
  }
  & ${CardText} {
    font-size: ${({ theme, styling = 'secondary' }) => theme.card[styling].textSizes};
  }
  ${({ styling }) =>
    styling === 'action'
      ? `&:hover {
        cursor: pointer;
        background-color: ${colors.greyLightest};

        & > ${CardLineItem}:last-of-type { 
          svg {
            color: ${colors.greyDark};
          }
        }
      } 
  }`
      : null}
`;

const Card = ({ children, ...props }: React.HTMLAttributes<HTMLDivElement> & CardProps) => {
  const theme = useThemeContext();
  return (
    <StyledCard theme={theme} {...props}>
      {children}
    </StyledCard>
  );
};

export default Card;
