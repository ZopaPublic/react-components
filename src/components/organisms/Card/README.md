### Summary

`<Card />` is used to group together content and actions.

It comes in a lot of variations to suit different needs.

### Example

- Full example of responsive cards

```tsx
import {
  Card,
  Button,
  FlexRow,
  FlexCol,
  FlexContainer,
  useViewport,
  breakpoints,
  grid,
} from '@zopauk/react-components';

<FlexContainer>
  <FlexRow>
    <FlexCol xs={12} l={4} className="mb-4">
      <Card layout={useViewport().width > grid.breakpoints.l ? 'vertical' : 'horizontal'}>
        <Card.Image overlayWith={<a href="http://zopa.com" target="_blank" />} url="http://placekitten.com/500/300" />
        <Card.Content>
          <Card.Heading>Macaroon bear claw powder</Card.Heading>
          <Card.Text>Ice cream marzipan marshmallow caramels sweet. Bonbon croissant lemon drops marzipan.</Card.Text>
          <Card.Actions>
            <Button styling="secondary" fullWidth={useViewport().width > grid.breakpoints.l}>
              Button
            </Button>
          </Card.Actions>
        </Card.Content>
      </Card>
    </FlexCol>
    <FlexCol xs={12} l={4} className="mb-4">
      <Card layout={useViewport().width > grid.breakpoints.l ? 'vertical' : 'horizontal'}>
        <Card.Image overlayWith={<a href="http://zopa.com" target="_blank" />} url="http://placekitten.com/500/300" />
        <Card.Content>
          <Card.Heading>Macaroon bear claw powder</Card.Heading>
          <Card.Text>Ice cream marzipan marshmallow caramels sweet. Bonbon croissant lemon drops marzipan.</Card.Text>
          <Card.Actions>
            <Button styling="secondary" fullWidth={useViewport().width > grid.breakpoints.l}>
              Button
            </Button>
          </Card.Actions>
        </Card.Content>
      </Card>
    </FlexCol>
    <FlexCol xs={12} l={4} className="mb-4">
      <Card layout={useViewport().width > grid.breakpoints.l ? 'vertical' : 'horizontal'}>
        <Card.Image overlayWith={<a href="http://zopa.com" target="_blank" />} url="http://placekitten.com/500/300" />
        <Card.Content>
          <Card.Heading>Macaroon bear claw powder</Card.Heading>
          <Card.Text>Ice cream marzipan marshmallow caramels sweet. Bonbon croissant lemon drops marzipan.</Card.Text>
          <Card.Actions>
            <Button styling="secondary" fullWidth={useViewport().width > grid.breakpoints.l}>
              Button
            </Button>
          </Card.Actions>
        </Card.Content>
      </Card>
    </FlexCol>
  </FlexRow>
</FlexContainer>;
```

- Vertical card with an image, heading, text, and a full width button

```tsx
import { Card, Button, FlexContainer, FlexCol, FlexRow } from '@zopauk/react-components';

<FlexContainer>
  <FlexRow>
    <FlexCol xs={4}>
      <Card styling="primary">
        <Card.Image overlayWith={<a href="http://zopa.com" target="_blank" />} url="http://placekitten.com/500/300" />
        <Card.Content>
          <Card.Heading>Primary card</Card.Heading>
          <Card.Text>Ice cream marzipan marshmallow caramels sweet. Bonbon croissant lemon drops marzipan.</Card.Text>
          <Card.Actions>
            <Button styling="primary" fullWidth>
              Button
            </Button>
          </Card.Actions>
        </Card.Content>
      </Card>
    </FlexCol>
    <FlexCol xs={4}>
      <Card>
        <Card.Image overlayWith={<a href="http://zopa.com" target="_blank" />} url="http://placekitten.com/500/300" />
        <Card.Content>
          <Card.Heading>Secondary card</Card.Heading>
          <Card.Text>Ice cream marzipan marshmallow caramels sweet. Bonbon croissant lemon drops marzipan.</Card.Text>
          <Card.Actions>
            <Button styling="secondary" fullWidth>
              Button
            </Button>
          </Card.Actions>
        </Card.Content>
      </Card>
    </FlexCol>
  </FlexRow>
</FlexContainer>;
```

- Vertical card with a heading, text, and a full width button

```tsx
import { Card, Button, FlexContainer, FlexRow, FlexCol } from '@zopauk/react-components';

<FlexContainer>
  <FlexRow>
    <FlexCol xs={4}>
      <Card styling="primary">
        <Card.Content>
          <Card.Heading>Primary card</Card.Heading>
          <Card.Text>Ice cream marzipan marshmallow caramels sweet. Bonbon croissant lemon drops marzipan.</Card.Text>
          <Card.Actions>
            <Button styling="primary" fullWidth>
              Button
            </Button>
          </Card.Actions>
        </Card.Content>
      </Card>
    </FlexCol>
    <FlexCol xs={4}>
      <Card>
        <Card.Content>
          <Card.Heading>Secondary card</Card.Heading>
          <Card.Text>Ice cream marzipan marshmallow caramels sweet. Bonbon croissant lemon drops marzipan.</Card.Text>
          <Card.Actions>
            <Button styling="secondary" fullWidth>
              Button
            </Button>
          </Card.Actions>
        </Card.Content>
      </Card>
    </FlexCol>
  </FlexRow>
</FlexContainer>;
```

- Vertical card with a heading, text, and a regular button

```tsx
import { Card, Button, FlexContainer, FlexRow, FlexCol } from '@zopauk/react-components';

<FlexContainer>
  <FlexRow>
    <FlexCol xs={4}>
      <Card styling="primary">
        <Card.Content>
          <Card.Heading>Primary card</Card.Heading>
          <Card.Text>Ice cream marzipan marshmallow caramels sweet. Bonbon croissant lemon drops marzipan.</Card.Text>
          <Card.Actions>
            <Button styling="primary">Button</Button>
          </Card.Actions>
        </Card.Content>
      </Card>
    </FlexCol>
    <FlexCol xs={4}>
      <Card>
        <Card.Content>
          <Card.Heading>Secondary card</Card.Heading>
          <Card.Text>Ice cream marzipan marshmallow caramels sweet. Bonbon croissant lemon drops marzipan.</Card.Text>
          <Card.Actions>
            <Button styling="secondary">Button</Button>
          </Card.Actions>
        </Card.Content>
      </Card>
    </FlexCol>
  </FlexRow>
</FlexContainer>;
```

- Vertical card with a heading, text, and two button

```tsx
import { Card, Button, FlexContainer, FlexRow, FlexCol } from '@zopauk/react-components';

<FlexContainer>
  <FlexRow>
    <FlexCol xs={4}>
      <Card styling="primary">
        <Card.Content>
          <Card.Heading>Primary card</Card.Heading>
          <Card.Text>Ice cream marzipan marshmallow caramels sweet. Bonbon croissant lemon drops marzipan.</Card.Text>
          <Card.Actions>
            <Button styling="secondary">Button</Button>
            <Button styling="primary">Button</Button>
          </Card.Actions>
        </Card.Content>
      </Card>
    </FlexCol>
    <FlexCol xs={4}>
      <Card>
        <Card.Content>
          <Card.Heading>Secondary card</Card.Heading>
          <Card.Text>Ice cream marzipan marshmallow caramels sweet. Bonbon croissant lemon drops marzipan.</Card.Text>
          <Card.Actions>
            <Button styling="secondary">Button</Button>
            <Button styling="primary">Button</Button>
          </Card.Actions>
        </Card.Content>
      </Card>
    </FlexCol>
  </FlexRow>
</FlexContainer>;
```

- Vertical secondary card with just a heading and text

```tsx
import { Card, Button, FlexContainer, FlexRow, FlexCol } from '@zopauk/react-components';

<FlexContainer>
  <FlexRow>
    <FlexCol xs={4}>
      <Card>
        <Card.Content>
          <Card.Heading>Secondary card</Card.Heading>
          <Card.Text>Ice cream marzipan marshmallow caramels sweet. Bonbon croissant lemon drops marzipan.</Card.Text>
        </Card.Content>
      </Card>
    </FlexCol>
  </FlexRow>
</FlexContainer>;
```

- Vertical brand card

```tsx
import { Card, Button, FlexContainer, FlexRow, FlexCol } from '@zopauk/react-components';

<FlexContainer>
  <FlexRow>
    <FlexCol xs={4}>
      <Card styling="brand">
        <Card.Content>
          <Card.Heading>Brand card</Card.Heading>
          <Card.Text>Ice cream marzipan marshmallow caramels sweet. Bonbon croissant lemon drops marzipan.</Card.Text>
        </Card.Content>
      </Card>
    </FlexCol>
  </FlexRow>
</FlexContainer>;
```

- Info card with icon, heading and text

`<Card.LineItem />` should be rendered outside of `<Card.Content />`

```tsx
import { Card, Button, FlexContainer, FlexRow, FlexCol, Icon } from '@zopauk/react-components';
import { faCalendarAlt } from '@fortawesome/free-solid-svg-icons';

<FlexContainer>
  <FlexRow>
    <FlexCol xs={4}>
      <Card styling="info" layout="horizontal" className="py-2">
        <Card.LineItem align="start">
          <Icon variant={faCalendarAlt} color={colors.grey} size="lg" className="ml-2" />
        </Card.LineItem>
        <Card.Content className="pl-2">
          <Card.Heading>Info card</Card.Heading>
          <Card.Text>Ice cream marzipan marshmallow caramels sweet. Bonbon croissant lemon drops marzipan.</Card.Text>
        </Card.Content>
      </Card>
    </FlexCol>
  </FlexRow>
</FlexContainer>;
```

- Horizontal action card with single line item

`<Card.LineItem />` should be rendered outside of `<Card.Content />`

```tsx
import { Card, Button, FlexContainer, FlexRow, FlexCol } from '@zopauk/react-components';
import Icon from '../../atoms/Icon/Icon.tsx';
import { faChevronRight } from '@fortawesome/free-solid-svg-icons';

<FlexContainer>
  <FlexRow>
    <FlexCol xs={4}>
      <Card styling="action" layout="horizontal">
        <Card.Content>
          <Card.Heading>Card with line item</Card.Heading>
          <Card.Text>Ice cream marzipan marshmallow caramels sweet. Bonbon croissant lemon drops marzipan.</Card.Text>
        </Card.Content>
        <Card.LineItem>
          <Icon variant={faChevronRight} color={colors.grey} />
        </Card.LineItem>
      </Card>
    </FlexCol>
  </FlexRow>
</FlexContainer>;
```

- Horizontal action card with two line items

`<Card.LineItem />` should be rendered outside of `<Card.Content />`

```tsx
import { Card, Button, FlexContainer, FlexRow, FlexCol } from '@zopauk/react-components';
import Icon from '../../atoms/Icon/Icon.tsx';
import { faPiggyBank, faChevronRight } from '@fortawesome/free-solid-svg-icons';

<FlexContainer>
  <FlexRow>
    <FlexCol xs={6}>
      <Card styling="action" layout="horizontal">
        <Card.LineItem>
          <Icon variant={faPiggyBank} color={colors.brand} size="2x" className="ml-4" />
        </Card.LineItem>
        <Card.Content className="py-6 pl-1">
          <Card.Heading className="mb-1">Card with additional line item</Card.Heading>
          <Card.Text>Ice cream marzipan marshmallow caramel piggy.</Card.Text>
        </Card.Content>
        <Card.LineItem>
          <Icon variant={faChevronRight} color={colors.grey} />
        </Card.LineItem>
      </Card>
    </FlexCol>
  </FlexRow>
</FlexContainer>;
```
