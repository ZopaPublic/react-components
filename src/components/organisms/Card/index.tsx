import React, { HTMLAttributes } from 'react';

import CardActions from './CardActions/CardActions';
import CardComponent, { CardProps } from './Card/Card';
import CardContent from './CardContent/CardContent';
import CardHeading from './CardHeading/CardHeading';
import CardImage from './CardImage/CardImage';
import CardText from './CardText/CardText';
import CardLineItem from './CardLineItem/CardLineItem';

const Card = (props: CardProps & HTMLAttributes<HTMLDivElement>) => <CardComponent {...props} />;

export default Object.assign(Card, {
  Actions: CardActions,
  Content: CardContent,
  Heading: CardHeading,
  Image: CardImage,
  Text: CardText,
  LineItem: CardLineItem,
});
