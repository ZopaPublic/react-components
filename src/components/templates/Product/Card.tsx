import React, { ReactNode } from 'react';
import styled from 'styled-components';
import classnames from 'classnames';

import { Card, breakpoints, Heading, Text, colors, useViewport } from '../../../index';

interface ProductTemplateCardProps {
  notification?: ReactNode;
  title?: ReactNode;
  subtitle?: ReactNode;
  className?: string;
  children: React.ReactNode;
}

const StyledCard = styled(Card)`
  @media (max-width: ${breakpoints.phone}px) {
    border-radius: 0;
  }
`;

function SizedHeading(props: { children: ReactNode }) {
  const { width = 0 } = useViewport();
  return width > breakpoints.phone ? (
    <Heading as="h2" size="h3">
      {props.children}
    </Heading>
  ) : (
    <Heading as="h2" size="h4">
      {props.children}
    </Heading>
  );
}

export function ProductTemplateCard(props: ProductTemplateCardProps) {
  return (
    <StyledCard className={classnames('px-4 py-6 m:p-8', props.className ? props.className : '')}>
      {props.notification ? props.notification : null}
      <SizedHeading>{props.title}</SizedHeading>
      {props.subtitle ? (
        <Text className="mt-2" color={colors.greyDark}>
          {props.subtitle}
        </Text>
      ) : null}
      {props.children}
    </StyledCard>
  );
}
