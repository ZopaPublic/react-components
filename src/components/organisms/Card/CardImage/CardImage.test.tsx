import React from 'react';
import { render } from '@testing-library/react';
import axe from '../../../../../axe-helper';
import Card from '..';

describe('<Card.Image />', () => {
  it('renders with a link', () => {
    const testId = 'card-image-link';
    const { queryByTestId } = render(
      <Card>
        <Card.Image overlayWith={<a href="http://www.zopa.com" data-automation={testId} />} url="" />
      </Card>,
    );
    expect(queryByTestId(testId)).toBeTruthy();
  });

  it('renders the component with no a11y violations', async () => {
    const { container } = render(
      <Card>
        <Card.Image url="" />
      </Card>,
    );
    const results = await axe(container.innerHTML);
    expect(container.firstChild).toMatchSnapshot();
    expect(results).toHaveNoViolations();
  });
});
