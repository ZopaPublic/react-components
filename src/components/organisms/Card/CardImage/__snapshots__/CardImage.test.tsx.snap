// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Card.Image /> renders the component with no a11y violations 1`] = `
.c2 {
  position: relative;
}

.c2 > a,
.c2 > button,
.c2 > div {
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
}

.c3 {
  height: 100%;
  width: 100%;
  background-image: url();
  background-size: cover;
  background-repeat: no-repeat;
  background-position: center;
}

.c0 {
  height: 100%;
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  overflow: hidden;
  background-color: #FFFFFF;
  border: 1px solid #EFEFEF;
  border-radius: 12px;
  box-shadow: 0 1px 0 0 #D4D7D9;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c0 > .c1 {
  -webkit-flex: 0 1 200px;
  -ms-flex: 0 1 200px;
  flex: 0 1 200px;
}

.c0 .zrc__card-heading {
  font-size: 16px;
}

<div
  class="c0"
>
  <div
    class="c1 c2"
  >
    <div
      class="c3"
    />
  </div>
</div>
`;
